thetis/timeintegrator.py

```diff
@@ -786,6 +1121,144 @@ def get_final_solution(self, output=None):
             output.assign(self.solution_old)
 
 
+class DIRKGenericNew(TimeIntegrator):
+    """
+    Generic implementation of Diagonally Implicit Runge Kutta schemes.
+
+    Method is defined by its Butcher tableau given as arguments
+
+    c[0] | a[0, 0]
+    c[1] | a[1, 0] a[1, 1]
+    c[2] | a[2, 0] a[2, 1] a[2, 2]
+    ------------------------------
+         | b[0]    b[1]    b[2]
+
+    This method also works for explicit RK schemes if one with the zeros on the first row of a.
+    """
+    def __init__(self, equation, solution, fields, dt, a, b, c,
+                 bnd_conditions=None, solver_parameters={}):
+        """
+        Create new DIRK solver.
+
+        Parameters
+        ----------
+        equation : equation object
+            the equation to solve
+        dt : float
+            time step (constant)
+        a  : array_like (n_stages, n_stages)
+            coefficients for the Butcher tableau, must be lower diagonal
+        b,c : array_like (n_stages,)
+            coefficients for the Butcher tableau
+        solver_parameters : dict
+            PETSc options for solver
+        terms_to_add : 'all' or list of 'implicit', 'explicit', 'source'
+            Defines which terms of the equation are to be added to this solver.
+            Default 'all' implies terms_to_add = ['implicit', 'explicit', 'source']
+        """
+        super(DIRKGenericNew, self).__init__(equation, solver_parameters)
+        self.solver_parameters.setdefault('snes_monitor', False)
+        self.solver_parameters.setdefault('snes_type', 'newtonls')
+
+        self.n_stages = len(b)
+        self.a = a
+        self.b = b
+        self.c = c
+
+        fs = self.equation.function_space
+        self.dt = dt
+        self.dt_const = Constant(dt)
+        self.solution_old = solution
+
+        test = self.equation.test
+
+        from firedrake.functionspaceimpl import MixedFunctionSpace, WithGeometry
+        mixed_space = (isinstance(fs, MixedFunctionSpace) or
+                       (isinstance(fs, WithGeometry) and
+                        isinstance(fs.topological, MixedFunctionSpace)))
```

Comment:
You can use `len(fs) > 1` instead for this check if you like.  Alternately, you can always use `fs.topological`, even if `fs` is not a `WithGeometry` (in which case you have `fs.topological is fs` returning `True`).